### ISSUE TITLE
Determinism in TensorRT-LLM

### ISSUE THREAD

**Issue by njaramish** [NVIDIA/TensorRT-LLM#2474]
Hello, I have a couple of determinism-related questions. Thank you very much in advance for your insight. At a high level, my goal is to get the highest-throughput deterministic model generations that I can, understanding that I will be reducing throughput in order to achieve determinism. 

1. I understand that TensorRT-LLM [engines will select different kernels based on batch size](https://github.com/NVIDIA/TensorRT-LLM/issues/1336), thus we expect input a given prompt `text` to have different generation logits when run at different batch sizes (e.g. `[text]` vs `[text, text, text]`). 
    1. I have been able to get deterministic generation by using the inflight batcher with max_batch_size=1, however this is highly inefficient since many of my generated sequences are long. 
    2. I was considering statically batching, however it seems like this will not be supported in the future given the deprecation of the V1 batcher. Given this, what is currently the "best" way to get high-throughput deterministic generations?   
2. I understand that TensorRT [engine builds can be non-deterministic](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#algorithm-select) due to system noise while the builder is profiling different kernels. I have not yet done rigorous testing, but I'm getting reproducible results on a 8xH100 server that is simultaneously running other workloads. Are there steps taken in TensorRT-LLM to ensure a deterministic engine build, or am I just luckily getting reproducible results? 
    1. Are there any flags I can set during the build stage that will encourage reproducible engine builds, even if a purely deterministic outcome cannot be guaranteed?

**Comment by Superjomn** [comment]
Hi @njaramish, sorry for the late reply. 
For emulating static batching, @kaiyux do you have a good practice?

> Are there steps taken in TensorRT-LLM to ensure a deterministic engine build, or am I just luckily getting reproducible results?

Achieving determinism across different engine builds isn't possible, but we recommend building on clean, empty nodes, which can help mitigate inconsistencies.

> Are there any flags 
`export FORCE_DETERMINISTIC=1` will help. But it may hurt performance, especially for the MoE.

**Comment by njaramish** [comment]
Thanks @Superjomn, I appreciate the insight! Regarding the engine build step, are engine builds expected to be non-deterministic even when using the profiler cache from a prior engine build?
